python/cudf/cudf/core/dataframe.py

```diff
@@ -4531,130 +4529,322 @@ def _sizeof_fmt(num, size_qualifier):
+               categorical
+        count            3
+        unique           3
+        top              d
+        freq             1
+
+        Excluding numeric columns from a ``DataFrame`` description.
+
+        >>> df.describe(exclude=[np.number])
+               categorical object
+        count            3      3
+        unique           3      3
+        top              d      a
+        freq             1      1
+
+        Excluding object columns from a ``DataFrame`` description.
+
+        >>> df.describe(exclude=[object])
+               categorical numeric
+        count            3     3.0
+        unique           3    <NA>
+        top              d    <NA>
+        freq             1    <NA>
+        mean          <NA>     2.0
+        std           <NA>     1.0
+        min           <NA>     1.0
+        25%           <NA>     1.5
+        50%           <NA>     2.0
+        75%           <NA>     2.5
+        max           <NA>     3.0
+        """
 
         if not include and not exclude:
-            numeric_data = self.select_dtypes(np.number)
-            output_frame = _create_output_frame(numeric_data, percentiles)
+            default_include = [np.number]
+            if datetime_is_numeric:
+                default_include.append("datetime")
+            data_to_describe = self.select_dtypes(include=default_include)
+            if len(data_to_describe.columns) == 0:
+                data_to_describe = self
 
         elif include == "all":
-            if exclude:
-                raise ValueError("Cannot exclude when include='all'.")
-
-            included_data = self.select_dtypes(np.number)
-            output_frame = _create_output_frame(included_data, percentiles)
-            logging.warning(
-                "Describe does not yet include StringColumns or "
-                "DatetimeColumns."
-            )
+            if exclude is not None:
+                raise ValueError("exclude must be None when include is 'all'")
 
+            data_to_describe = self
         else:
-            if not include:
-                include = np.number
-
-            included_data = self.select_dtypes(
+            data_to_describe = self.select_dtypes(
                 include=include, exclude=exclude
             )
-            if included_data.empty:
+
+            if data_to_describe.empty:
                 raise ValueError("No data of included types.")
-            output_frame = _create_output_frame(included_data, percentiles)
 
-        return output_frame
+        describe_series = [
+            data_to_describe[col].describe(percentiles=percentiles)
+            for col in data_to_describe.columns
+        ]
+        names = []
+        ldesc_indexes = sorted((x.index for x in describe_series), key=len)
+        for idxnames in ldesc_indexes:
+            for name in idxnames.to_pandas():
+                if name not in names:
+                    names.append(name)
```

Comment:
Instead of using a list for `names` here we'd do better with something like an `OrderedDict` that allows for O(1) lookups so we can always do `names[name] = None`.

Reply:
Can also turn this into a comprehension 😄